src/Std/Time/Zoned/ZonedDateTime.lean

```diff
@@ -0,0 +1,321 @@
+/-
+Copyright (c) 2024 Lean FRO, LLC. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Sofia Rodrigues
+-/
+prelude
+import Std.Time.Zoned.DateTime
+import Std.Time.Zoned.ZoneRules
+
+namespace Std
+namespace Time
+
+set_option linter.all true
+
+/--
+An existential version of `DateTime` that encapsulates a `DateTime` value without explicitly storing
+the time zone.
+-/
+def ZonedDateTime := Sigma DateTime
+
+instance : BEq ZonedDateTime where
+  beq x y := x.fst == y.fst && x.snd.timestamp == y.snd.timestamp
+
+instance : CoeDep ZonedDateTime d (DateTime d.fst) where
+  coe := d.snd
+
+instance : Inhabited ZonedDateTime where
+  default := ⟨Inhabited.default, Inhabited.default⟩
+
+namespace ZonedDateTime
+open DateTime
+
+/--
+Creates a new `ZonedDateTime` out of a `Timestamp`
+-/
+@[inline]
+def ofTimestamp (tm : Timestamp) (tz : TimeZone) : ZonedDateTime :=
+  ⟨tz, DateTime.ofTimestamp tm tz⟩
+
+/--
+Creates a new `Timestamp` out of a `ZonedDateTime`
+-/
+@[inline]
+def toTimestamp (date : ZonedDateTime) : Timestamp :=
+  date.snd.toTimestamp
+
+/--
+Creates a new `ZonedDateTime` out of a `Timestamp`
+-/
+@[inline]
+def ofZoneRules (tm : Timestamp) (rules : TimeZone.ZoneRules) : Option ZonedDateTime := do
+  let transition ← rules.findTransitionForTimestamp tm
+  let tm := rules.applyLeapSeconds tm
+  return ofTimestamp tm transition.localTimeType.getTimeZone
+
```

Comment:
There seems to be an important function missing here, namely one that takes a `PlainDateTime` and a `TimeZone.ZoneRules` and returns an `Option ZonedDateTime` representing the time assuming the `PlainDateTime` was local for the zone rules.